### DISCUSSION TITLE
Error while loading shared libraries: libnss3.so: cannot open shared object file: No such file or directory

### DISCUSSION THREAD

**Discussion by yazeedsawalha** [vercel/community#3545] [Help]
### Summary

In the process of upgrading my Puppeteer project to Node.js 18.x as required by Vercel, I got missing libs libnss3.so

### Steps to Reproduce

The situation arose when I needed to upgrade my Puppeteer project to Node.js 18.x, a requirement set by Vercel. As I made the necessary upgrades and prepared to run the project, I encountered an error during runtime that caught my attention, it's work fine on node js 14. The error message reads as follows:

`Failed to launch the browser process!\n/tmp/chromium: error while loading shared libraries: libnss3.so: cannot open shared object file: No such file or directory\n\n\nTROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md\n"`

The error message indicated that the issue revolved around launching the browser process, specifically related to a missing shared library named libnss3.so. It became clear that resolving this error would involve ensuring the proper dependencies were present in my system, and it's worth noting that libnss3 is a crucial component of the libnss3 package

**Comment by AaronMcCloskey** [comment]
Also having this issue, I have seen a lot of solutions online to downgrade to node 14, but that's no longer possible on vercel. (and would mean downgrading a few of my packages requiring node 16+)

Tried downgrading `chrome-aws-lambda` & `puppeteer-core`, but with no luck.

**Comment by amyegan** [comment]
Sorry for the delay. Could you try upgrading your Puppeteer dependencies to the latest versions since your project is using Node 18?

**Comment by jaxball** [comment]
+1 to this issue. I'm going off this [virtual-event-starter-kit](https://github.com/vercel/virtual-event-starter-kit/blob/main/lib/screenshot.ts) example from vercel. 

```
import chrome from 'chrome-aws-lambda';
import puppeteer from 'puppeteer-core';

export default async function screenshot(url: string) {
  const options = process.env.AWS_REGION
    ? {
        args: chrome.args,
        executablePath: await chrome.executablePath,
        headless: chrome.headless
      }
    : {
        args: [],
        executablePath:
          process.platform === 'win32'
            ? 'C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe'
            : process.platform === 'linux'
            ? '/usr/bin/google-chrome'
            : '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome'
      };
  const browser = await puppeteer.launch(options);
```

npm versions: 
```
"puppeteer-core": "^10.5.0",
"chrome-aws-lambda": "^10.1.0", (latest version, updated 2 years ago).
```
Node.js 18.x runtime deployed on vercel.

@amyegan would be great to get a solution to this.

**Comment by AliAryanTech** [comment]
Well, it seems like we don't get a solution to this problem yet

**Comment by elvislicsari** [comment]
Not necessarily a solution but rather a workaround, which is good enough in my case.
package.json
```
"dependencies": {
    "express": "^4.18.2",
    "puppeteer": "^21.6.1",
    "puppeteer-core": "^21.6.1"
  },
"engines": {
  "node": "18.x"
}
```
and then, in your app.js file
```
import puppeteer from 'puppeteer-core'
...
const browser = await puppeteer.connect({browserWSEndpoint: 'wss://chrome.browserless.io?token=YOUR_TOKEN_HERE'})
```

Basically you replace the whole chromium installation issues with this [browserless.io](https://www.browserless.io/) service. 
Go to https://www.browserless.io/, create an account, choose a plan (free plan works for my use case), get the token from your account and it should work on local & vercel.

**Comment by arifluthfi16** [comment]
any update on this??

**Comment by hengliu0919** [comment]
any update on this?

**Comment by onurcagan** [comment]
Rip, also had to update my project that was using node 14 and was working just fine and now I can no longer get it to work, are there any works on fixing this?

**Comment by AliAryanTech** [comment]
alright, that's enough guys! 
I have figured out the solution for this problem.

package.json:
```json
{
   "private": true,
   "dependencies": {
      "@sparticuz/chromium": "^119.0.2",
      "puppeteer-core": "^21.5.1"
   }
}
```
Here's the code:
```js
const chrome = require('@sparticuz/chromium');
const puppeteer = require('puppeteer-core');
const production = process.env.NODE_ENV === 'production';

const getPage = async () => {
    const browser = await puppeteer.launch(
        production ? {
            args: chrome.args,
            defaultViewport: chrome.defaultViewport,
            executablePath: await chrome.executablePath(),
            headless: 'new',
            ignoreHTTPSErrors: true
        } : {
            headless: 'new',
            executablePath: '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome'
        }
    );
    const page = await browser.newPage();
    return page;
};
```
that's how you can do your thing with node 18x, thanks me later! 🫠

**Reply by onurcagan** [reply]
Yep, this is the solution I've found elsewhere as well after commenting here yesterday and was going to comment later today. Thanks for the comment @AliAryanTech 

And realized why I hate node and js environment with the constant need of having to hack stuff together and hope for the best while trying to abide by the rules of providers...

**Reply by AliAryanTech** [reply]
Hahaha, that's true tho
